tests/SchemaTest.php

```diff
@@ -69,28 +76,82 @@ private function extractSchemas(Finder $files) : iterable
         foreach ($files as $file) {
             $dom = FluentDOM::load($file->getContents());
 
-            $xmlModel = $dom('substring-before(substring-after(/processing-instruction("xml-model"), \'"\'), \'"\')');
-            $schema = "{$file->getPath()}/{$xmlModel}";
+            yield $file->getRelativePathname() => [
+                $dom,
+                $this->findValidator($dom, $file),
+                $this->findExpectedFailures($dom, $file),
+            ];
+        }
+    }
 
-            $validator = new RelaxNgValidator($schema);
+    private function findValidator(Document $dom, SplFileInfo $file) : XmlValidator
+    {
+        $validators = map(
+            $dom('/processing-instruction("xml-model")'),
+            function (ProcessingInstruction $instruction) use ($file) {
+                $parsed = $this->parseProcessingInstruction($instruction, $file);
+                $schema = "{$file->getPath()}/{$parsed['href']}";
+                if (!is_readable($schema)) {
+                    throw new LogicException("Failed to read schema {$schema} in {$file->getRelativePathname()}");
+                }
+                switch ($parsed['schematypens']) {
+                    case 'http://relaxng.org/ns/structure/1.0':
+                        return new RelaxNgValidator($schema);
+                    case 'http://purl.oclc.org/dsdl/schematron':
+                        return new SchematronValidator($schema);
```

Comment:
Would have handy to provide this in the XML Validator library (a validator that reads the processing instructions).

Reply:
It is highly coupled to that library anyway, so it could be pushed upstream. It is one use case of many as people can still use the underlying validators.